untidy/contaminators.py

```diff
@@ -3,129 +3,197 @@
+    num_contaminated = int(prop_contaminated * num_obs)
+
+    if isinstance(data, pd.DataFrame):
+        # Find columns to sample from
+        if col_type == "any":
+            cols_to_sample = range(len(data.columns))
+        elif col_type.startswith("str"):
+            cols_to_sample = [
+                list(data.columns).index(col)
+                for col in data.select_dtypes(include=["object", "category"]).columns
+            ]
+        elif col_type.startswith("num"):
+            cols_to_sample = [
+                list(data.columns).index(col)
+                for col in data.select_dtypes(include=["float64", "int64"]).columns
+            ]
+        cols_to_sample = list(cols_to_sample)
+        sampled_col_idx = random.choices(cols_to_sample, k=num_contaminated)
+        sampled_row_idx = random.choices(list(range(len(data))), k=num_contaminated)
+        idx = list(set(zip(sampled_row_idx, sampled_col_idx)))
+    elif isinstance(data, pd.Series):
+        idx = list(data.sample(n=num_contaminated).index)
     else:
-        raise TypeError('data should be pd.Series or pd.DataFrame')
+        raise TypeError("data should be pd.Series or pd.DataFrame")
+
+    return idx
 
-    return num_rows, num_cols
 
-def get_random_indices(num_rows, num_cols, corruption_level = 4):
+""" Functions to contaminate text columns """
+
+
+def add_noise_to_strings(clean_data, corruption_level=4):
     """
-    Get random indeces to contaminate
+    Introduce noise to strings in clean data
 
     Parameters
     ----------
-    num_rows: int
-        number of rows in data to contaminate
-    num_cols: int
-        number of columns in data to contaminate
-    corruption_level int, optional
+    clean_data: pd.Series or pd.DataFrame
+        data to be contaminated with superfluous characters
+    corruption_level: int, optional
         level of corruption, should be between 0 and 10, where 0 leaves the dataset as is, 10
         is the highest level of contamination
 
     Returns
     -------
-    list of indeces to be contaminated
+    data: pd.DataFrame or pd.Series
+        contaminated dataset
     """
+    data = clean_data.copy()
 
-    # define amount of missing values to introduce
-    contamination_fraction = round(num_rows * num_cols * corruption_level/10 * 0.6)
+    # Find data cells to contaminate
+    idxs_to_contaminate = get_random_indices(
+        data, col_type="str", corruption_level=corruption_level
+    )
 
-    # generate randomized missing values indeces
-    
-    if num_cols>1:
-        cont_idxs = zip(random.choices(range(num_rows), k=contamination_fraction),
-                    random.choices(range(num_cols), k=contamination_fraction))
-    else:
-        cont_idxs = zip(random.choices(range(num_rows), k=contamination_fraction),
-                    [0 for i in range(contamination_fraction)])
-
-    return cont_idxs
+    # Perform contamination
+    noise_chars = "%&$?!#"
+    whitespace = "  "
```

Comment:
couldn't a white space be included in noise_chars so we wouldn't need whitespace? the random.choice later should then take care of randomly picking whitespace as well. We could use the parameter 'p' in random.choice to give more probability to the whitespace